datastore/fsds.nim

```diff
@@ -216,6 +216,10 @@ method query*(
     return success (key.some, data)
 
   iter.next = next
+  iter.dispose = proc(): Future[?!void] {.async.} =
```

Comment:
I think that disposing an inter should follow iter being finished, not the other way around.

Reply:
I'm sorry, I'm not sure what you mean. :o

Reply:
So setting `finished` to true, should be an internal event that's signaling that all items were yielded (line 197).

With this change, flag `finished` would be set to true whenever `dispose()`  would be called (not when all items were yielded). 

As a client of this API you can finish iteration prematurely and then call `dispose()`, but that would not mean that iterator yielded all elements.

Reply:
:red_circle: Well the client won't even finish iteration because `next` will still happily return the next element as it doesn't check the `finished` flag to decide what to do. So if you're relying on it returning `success (Key.none, ...)` to know when to stop, this will not have the intended effect.

I suppose that to get the right semantics (assuming this is the right one) you need to check the state of the finished flag in addition to the state of the walker iterator (e.g `if finished(walker) or iter.finished`).

Reply:
I think I understand the desired behavior and I'm gonne write a test for it.

Reply:
Update: Finished iters will always return None when asked for next. There's a test to ensure this, and ensure that disposed iters become finished.

Reply:
Not sure why we're (incorrectly) changing the semantics of the iterators now? 

as @tbekas is saying:

> So setting finished to true, should be an internal to iterator event that's signaling that all items were yielded (line 197).

Reply:
Yeah OK the concern I had is fixed by this.

Reply:
Yes, we can't assume the iterator will always run through all elements. I've added a test to show that you can properly clean up an iterator early by calling dispose on it. Any nexts would then yield nothing and the underlying resources can be closed.